File: java/720_Longest_Word in_Dictionary.java
/*
https://leetcode.com/problems/longest-word-in-dictionary/
Given a list of strings words representing an English Dictionary, find the longest word in words that can be built one
character at a time by other words in words. If there is more than one possible answer, return the longest word with
the smallest lexicographical order. If there is no answer, return the empty string.

Convert words to a set. For each word, if it is better than current longest then update longest if the set contains
all prefixes of the word.
Time - O(n**2), for each string examine possible all other strings
Space - O(n)
*/

class Solution {
    public String longestWord(String[] words) {

        String longest = "";
        Set<String> wordSet = new HashSet(Arrays.asList(words));

        for (String candidate:wordSet) {
            // candidate is longer or same length and lexicographically before longest
            if (candidate.length() > longest.length() || (candidate.length() == longest.length() && candidate.compareTo(longest) < 0)) {
                // check if all prefixes are in wordSet
                boolean allPrefixes = true;
                for (int i = 1; i <= candidate.length(); ++i) {
                    if (!wordSet.contains(candidate.substring(0, i))) {
                        allPrefixes = false;
                        break;
                    }
                }
                if (allPrefixes)
                    longest = candidate;
            }
        }

        return longest;
    }
}

File: python_1_to_1000/720_Longest_Word_in_Dictionary.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/longest-word-in-dictionary/
# Given a list of strings words representing an English Dictionary, find the longest word in words that can be built
# one character at a time by other words in words. If there is more than one possible answer, return the longest word
# with the smallest lexicographical order.
# If there is no answer, return the empty string.

# Map lengths of words to sets of words with the same length.
# For each word longer than the current candidates (initially only the empty string is a candidate), add word to
# next_candidates if its prefix without final letter is already a candidate.
# Alternatively, for each word better than the longest word already found, check all prefixes in set of dictionary.
# Time - O(nk) for n words of maximum length k
# Space - O(nk)

from collections import defaultdict

class Solution(object):
    def longestWord(self, words):
        """
        :type words: List[str]
        :rtype: str
        """
        length_to_words = defaultdict(set)

        for word in words:      # map word lengths to set of words
            length_to_words[len(word)].add(word)

        candidates = {""}       # initial candidate is the empty string
        length = 0              # length of current candidates

        while True:

            next_candidates = set()

            for longer_word in length_to_words[length + 1]: # check if each longer word can be built from any candidate
                if longer_word[:-1] in candidates:
                    next_candidates.add(longer_word)

            if not next_candidates:
                return sorted(list(candidates))[0]          # sort to return lexicographically lowest

            length += 1
            candidates = next_candidates